---
title: "Newtech - Metagenomics on milk (Spallanzani + CREA-ZA)"
author: "Nelson Nazzicari"
date: "`r Sys.Date()`"
output: 
  html_document:
    css: CREA.css
    toc: true
    toc_float: true
    number_sections: true
---

```{r setup, echo=FALSE, include=FALSE, message=FALSE}
library(knitr)
options(knitr.kable.NA = ' ', digits = 3) #this is for kable
library(ggplot2)
library(plyr)
library(printr)
library(DT)
library(xfun)
library(vegan)
library(reshape2)

xfun::pkg_load2(c('base64enc', 'htmltools', 'mime'))

source('~/research/CREA-ZA_Spallanzani_metabarcoding-grana-veg/script/data_manager.R')
infolder = '~/research/CREA-ZA_Spallanzani_metabarcoding-grana-veg/data/private/delivery_20200706/'
delivery_20200706 = load_all_OTU_files(infolder)
infolder = '~/research/CREA-ZA_Spallanzani_metabarcoding-grana-veg/data/private/delivery_20201009/'
delivery_20201009 = load_all_OTU_files(infolder)
raw_OTUs = c(delivery_20200706, delivery_20201009)

```

```{r cheatsheet, echo=FALSE, include=FALSE, eval=FALSE}
# EVALUATE CODE?  eval=TRUE
# PRINT CODE?     echo=TRUE
# PRINT OUTPUT?   include=TRUE
# UNWANTED STUFF? message=FALSE warning=FALSE error=FALSE

#attach a file (e.g. an xls)
xfun::embed_file('some_file.xls')

#data table with fancy selectors and searches
DT::datatable(cars)

#more humble data table
kable(cars, caption = "my cars")
```

```{r echo=FALSE, include=TRUE}
#this is your standard R chunk
```

# Document history

<!--
To highlight changes use
<span class="new"> your stuff </span>
-->

* 2020-11-04 - initial report
* 2020-11-09 - added section on bovine genome
* 2020-11-16 - added section on filtering, control samples, a first pass for milk samples
* 2020-11-16 - answer to open question 3
* 2020-11-17 - added heatmaps for frequencies at various taxonomic levels
* 2020-11-30 - added diversity analysis, alpha
* 2020-12-01 - added genus level for control samples, added reference to duplicate samples notes
* 2020-12-07 - added beta diversity, rarefaction curves, open question about diary homogeneity

# Aim

Metabarcoding analysis on Grana Padano milk, focusing on vegetable presence.

# Available samples

```{r echo=FALSE, include=TRUE}
samples = read.csv('~/research/CREA-ZA_Spallanzani_metabarcoding-grana-veg/data/private/Transcodifica Campioni_cleaned.csv', stringsAsFactors = FALSE, row.names = 1)
samples = samples[,c("Id_Lab", "type", "caseificio", "MESE", "data_prelievo")]
samples$province = sapply(samples$caseificio, FUN = substr, start=1, stop=2)
samples[samples$province == '', 'province'] = '-'
```

## Notes on samples

Here's a piece of correspondence with Anna and Rossana, for future reference:

---

* il campione con id_sample 10, 17 e 21 (ID_lab:177), è sempre lo stesso campione ma:
  * 10 e 17 due aliquote (per motivi di metodo estrattivo) della stessa estrazione ma con curve al nanodrop diverse;
  * 21 stesso campione di latte ma estratto con scala small (cioè men volume per un tentativo iniziale di miniaturizzare)
* campione  id_sample  9 e 22 (ID_lab:387): stesso campione di latte ma estratto con scala small (cioè men volume per un tentativo iniziale di miniaturizzare)


## By type

```{r echo=FALSE, include=TRUE}
samples_summary = ddply(samples, .(type), function(x){
  return(data.frame(
    samples = nrow(x),
    Id_Lab = paste(collapse = ', ', x$Id_Lab)
  ))
})

kable(samples_summary)
```

where "broken" indicates a non viable sample which was then re-sent and sequenced, "ctrl" indicate pure vegetable samples (e.g. fodder), "grana" and "non-grana" indicates milk samples from inside and outside Grana Padano areas, respectively.

## By province

```{r echo=FALSE, include=TRUE}
#from here on we don't need the broken samples anymore
samples = subset(samples, type != 'broken')

samples_summary = ddply(samples, .(province), function(x){
  return(data.frame(
    samples = nrow(x),
    Id_Lab = paste(collapse = ', ', x$Id_Lab)
  ))
})

kable(samples_summary)
```

## By extraction method

Two extraction methods have been tested but at the end we will only work with phenol.

# Fraction of Bovine genome

After a mail exchange with Slobodanka Radovic I obtained the OTU table for the second delivery, containing the numerosities for 27 samples, plus the sequences for each of the 4811 centroids analyzed.

A first analysis of the centroid cumulative numerosity led to the following plot:

```{r echo=FALSE, include=TRUE, eval=TRUE}

#read the OTU table
OTU = read.csv('~/research/CREA-ZA_Spallanzani_metabarcoding-grana-veg/data/private/otu_table-2.csv', stringsAsFactors = FALSE, row.names = 1)
OTU$taxonomy = NULL

#getting the most present centroids
centroids_cnt = rowSums(OTU)
OTU = OTU[order(centroids_cnt, decreasing = TRUE),]

#cumulative plot
centroids_cum = centroids_cnt
for (i in 1:length(centroids_cnt)){
  centroids_cum[i] = sum(centroids_cnt[1:i])
}
centroids_cum = centroids_cum / sum(centroids_cnt)

plot(centroids_cum, xlab = 'Centroids, sorted by numerosity', ylab = 'Cumulative reads, relative')
```

For reference, the 10 most numerous centroids account for the `r centroids_cum[10] * 100` % of all the reads.

To investigate what fraction of the available reads have bovine origin I've blasted the centroid sequences on the 
*Bos Taurus* genome via the [NCBI website](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch&PROG_DEF=blastn&BLAST_PROG_DEF=megaBlast&BLAST_SPEC=OGP__9913__10708).

```{r echo=FALSE, include=TRUE, eval=TRUE}
al = rbind(
  read.csv('~/research/CREA-ZA_Spallanzani_metabarcoding-grana-veg/data/private/centroids/lines1-5000_Alignment-HitTable.csv', stringsAsFactors = FALSE, header = FALSE),
  read.csv('~/research/CREA-ZA_Spallanzani_metabarcoding-grana-veg/data/private/centroids/lines5001-end_Alignment-HitTable.csv', stringsAsFactors = FALSE, header = FALSE)
)

#giving column names, for what we are interested in
colnames(al) = c('id1', 'id2', 'perc_id', 'match_length', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'e_value', 'bit_score')

#removing duplicates
al = ddply(al, .(id1), function(x){
  i = which.min(x$e_value)
  return(x[i,])
})

#filtering
al = subset(al, perc_id>0.95 & e_value < 0.00001 )

#how many of the reads are bovine
bovine_reads = sum(centroids_cnt[al$id1])

```
The sequencing results are filtered with the following criteria:

* percent identity > 0.95
* e-value < 1e-5
* in case of multiple alignments I keep the one with smallest e_value

After the filtering `r nrow(al)` centroids align on the bovine genome, representing the `r 100*nrow(al)/length(centroids_cnt)` % of the total `r length(centroids_cnt)` centroids. In terms of reads, the `r 100*bovine_reads/sum(centroids_cnt)` % of reads are bovine.

## OPEN QUESTION 1: what fraction of unassigned data is bovine, exactly?
So far I've done a high level analysis of centroids, but it would be better to operate sample-by-sample.

# Filtering and conventions

This filtering strategy is taken straighforward from the [published paper on metabarcoding in grana padano](https://www.sciencedirect.com/science/article/pii/S0740002020302021?casa_token=pwjIEIBIUI0AAAAA:UAhVfUNTpp9OOG_iTAWkdc9GKUiy4cd1Ho2On19p80QX8MZHb-MCbDbnFn3a66D3DTn8XqerzTk). We may want to review these values at a later time. The current plan is:

* entries having less than 5 reads are removed and binned to unassigned (UN)
* entries (species?) representing 1% of the reads of a sample will be labeled "subdominant" and will be considered less important in the analysis

The taxonomic hierarchy we used is:

* reads
* domain
* clade/phylum
* class
* order
* family
* subfamily
* tribe
* subtribe
* genus
* species

with some levels (e.g. clade, subtribe) often absent.

# Control samples

In this section I focus on the three control samples: GT ("MANGIME X GP"), N ("NUCLEO"), S ("SOIA"), with the following expected compositions:

* S: soy (glycine max)
* GT: soy (48%), girasole (36%), "tritello" (triticum wheat), maize, "cruscami" (???, low percentage anyway)
* N: to be investigated

```{r echo=FALSE, include=TRUE}
level = 'domain'
OTU_ctrl = build_OTU_table(raw_OTUs, level) 

#subset to control
sel = OTU_ctrl$meta$type == 'ctrl'
OTU_ctrl = filter_OTU(OTU_ctrl, sel)

#removing clumsy code column
OTU_ctrl$meta$code = NULL

kable(OTU_ctrl$meta, row.names = FALSE)
```

Taking a look at the assigned **domains** we get:

```{r echo=FALSE, include=TRUE}
kable(OTU_ctrl$OTU_rel_screen, row.names = TRUE, col.names = paste(colnames(OTU_ctrl$OTU_rel), '%'))
```

Where "UN" means unassigned reads. All in all most of the reads are assigned, as expected in control samples. If we go deeper into the taxonomic tree, at **family** level we find (click on column names to reorder):

```{r echo=FALSE, include=TRUE}
level = 'family'
OTU_ctrl = build_OTU_table(raw_OTUs, level) 

#subset to control
sel = OTU_ctrl$meta$type == 'ctrl'
OTU_ctrl = filter_OTU(OTU_ctrl, sel)

#transpose, for ease of reading
df = data.frame(t(OTU_ctrl$OTU_rel_screen))
colnames(df) = OTU_ctrl$meta$Id_Lab
DT::datatable(df)
```

This table still is quite usable, but shows a trend: the number of unassigned entries grows when we go deeper into the
taxonomy tree. We already have about 8% of unassigned reads for GT and N. Going down to **genus** level:

```{r echo=FALSE, include=TRUE}
level = 'genus'
OTU_ctrl = build_OTU_table(raw_OTUs, level) 

#subset to control
sel = OTU_ctrl$meta$type == 'ctrl'
OTU_ctrl = filter_OTU(OTU_ctrl, sel)

#transpose, for ease of reading
df = data.frame(t(OTU_ctrl$OTU_rel_screen))
colnames(df) = OTU_ctrl$meta$Id_Lab
DT::datatable(df)
```

And to **species** level:

```{r echo=FALSE, include=TRUE}
level = 'species'
OTU_ctrl = build_OTU_table(raw_OTUs, level) 

#subset to control
sel = OTU_ctrl$meta$type == 'ctrl'
OTU_ctrl = filter_OTU(OTU_ctrl, sel)

#transpose, for ease of reading
df = data.frame(t(OTU_ctrl$OTU_rel_screen))
colnames(df) = OTU_ctrl$meta$Id_Lab
DT::datatable(df)
```

This is to show that we may need to operate at a level different from **species**, or risk losing a lot of information.

## OPEN QUESTION 2: what taxonomic level?
How to decide what taxonomic level to use? I've read a few metabarcoding papers that use phylum/family. Would it be ok? Further data exploration will probably give us more hints.

# Milk samples - reads and vegetable portion

The first step is to take a look to **domain** level for milk samples. In this table I report all the metadata (province, date and so forth). I then have three data columns:

* Eukaryota: percentage of reads, for each sample, classified as Eukaryota
* UN : as above, but for unassigned reads
* Reads: absolute number of reads for that sample

Here's the table, sorted by decreasing number of unassigned:

```{r echo=FALSE, include=TRUE}
level = 'domain'
OTU_milk = build_OTU_table(raw_OTUs, level) 

#subset to non-control
sel = OTU_milk$meta$type != 'ctrl'
OTU_milk = filter_OTU(OTU_milk, sel)

#building a single table, showable
OTU_milk$merged = cbind(OTU_milk$meta, OTU_milk$OTU_rel_screen)
OTU_milk$merged$code = NULL
OTU_milk$merged$reads = rowSums(OTU_milk$OTU, na.rm = TRUE)

#sorting by interesting col
OTU_milk$merged = OTU_milk$merged[order(OTU_milk$merged$UN, decreasing = TRUE),]

DT::datatable(OTU_milk$merged, rownames = FALSE, options = list(
  pageLength = 100))
```

A few notes:

* sample ID1701-1 is 100% unassigned. Also, it has a very low number of reads. I would consider this a failed sample?
    * **YES**
* for the other samples: assigned fractions go from just above 1% (sample ID1701-10) to 36.6% (sample ID1701-14)
* of the four non-grana, three (montagna, puglia, Na) have relatively high assignments (in the 8-10% range) and only one (sicilia) is in the low range (1.8%)
* looking at the distribution of reads, apart from the already cited low-reads sample, there are a few extremes (this is better appreciated sorting the table by number of reads):
    * sample ID1701-14 is the second-lowest in terms of reads, with 37k. The next would be sample ID1701-18, with 70k -  almost two times that amount
    * number of reads grows progressively, up until sample ID1701-25, which has 1.4M reads, almost four times the second most numerous sample (ID1557-21, with 400k reads)

## [CLOSED] OPEN QUESTION 3: are these reads levels ok?

I think I need to check with Slobodanka again, at least to confirm that this is expected.

**ANSWER**: it appears that the reads levels are acceptable, she suggests subsampling/oversampling reads if we need to normalize the samples (we don't, we'll use frequencies).


# Milk, heatmaps

```{r echo=FALSE, include=TRUE}
#declaring the breaks and labels for counting UN fraction at each taxonomic level
UN_breaks = c(0, 0.25, 0.50, 0.75, 0.8, 0.85, 0.9, 0.95, 1)
UN_labels = c('<25%', '25%-50%', '50%-75%', '75%-80%', '80%-85%', '85%-90%', '90%-95%', '>95%')
```


Several heatmaps for dominant (>1%) centroids, at various taxonomic levels.


## Clade/Phylum
```{r, fig.height = 8, fig.align = "center", echo=FALSE, include=TRUE}
level = 'clade_phylum'
OTU_milk = build_OTU_table(raw_OTUs, level, min_sample_reads = 0) 

#subset to non-control
sel = OTU_milk$meta$type != 'ctrl'
OTU_milk = filter_OTU(OTU_milk, sel)

#do heatmap
h = prepare_heatmap(OTU_milk$OTU_rel_noUN, min_abundance=0.01)

#show the map
print(h$gg)
```

At Phylum level 100% of the reads are Streptophyta, with the single exception of sample ID1707-1, which is 100% unclassified. This sample will be removed from next analyses.

Taking a look at the distribution of UN fractions:

```{r echo=FALSE, include=TRUE}
#counting UN fractions
df = data.frame(
  UN_fraction = UN_labels,
  samples = hist(OTU_milk$OTU_rel$UN, plot=FALSE, breaks = UN_breaks)$counts
)
  
#showing the table
kable(df)
```

## Family
```{r, fig.height = 8, fig.align = "center", echo=FALSE, include=TRUE}
level = 'family'
OTU_milk = build_OTU_table(raw_OTUs, level) 

#subset to non-control and remove bad sample
sel = OTU_milk$meta$type != 'ctrl'
OTU_milk = filter_OTU(OTU_milk, sel)

#do heatmap
h = prepare_heatmap(OTU_milk$OTU_rel_noUN, min_abundance = 0.01)

#show the map
print(h$gg)


#counting UN fractions
df = data.frame(
  UN_fraction = UN_labels,
  samples = hist(OTU_milk$OTU_rel$UN, plot=FALSE, breaks = UN_breaks)$counts
)
  
#showing the table
kable(df)
```

## Genus
```{r, fig.height = 8, fig.align = "center", echo=FALSE, include=TRUE}
level = 'genus'
OTU_milk = build_OTU_table(raw_OTUs, level) 

#subset to non-control and remove bad sample
sel = OTU_milk$meta$type != 'ctrl' 
OTU_milk = filter_OTU(OTU_milk, sel)

#do heatmap
h = prepare_heatmap(OTU_milk$OTU_rel_noUN, min_abundance = 0.01)

#show the map
print(h$gg)


#counting UN fractions
df = data.frame(
  UN_fraction = UN_labels,
  samples = hist(OTU_milk$OTU_rel$UN, plot=FALSE, breaks = UN_breaks)$counts
)
  
#showing the table
kable(df)
```


## Species
```{r, fig.height = 8, fig.align = "center", echo=FALSE, include=TRUE}
level = 'species'
OTU_milk = build_OTU_table(raw_OTUs, level) 

#subset to non-control and remove bad sample
sel = OTU_milk$meta$type != 'ctrl' 
OTU_milk = filter_OTU(OTU_milk, sel)

#do heatmap
h = prepare_heatmap(OTU_milk$OTU_rel_noUN, min_abundance = 0.01)

#show the map
print(h$gg)

#counting UN fractions
df = data.frame(
  UN_fraction = UN_labels,
  samples = hist(OTU_milk$OTU_rel$UN, plot=FALSE, breaks = UN_breaks)$counts
)
  
#showing the table
kable(df)
```

# Diversity analysis

There are many ways to quantify complexity with the aim of comparing communities objectively. The two main categories of methods are known as alpha diversity and beta diversity [Whittaker, 1960]. Alpha diversity measures the diversity within a single sample and is based on the number and relative abundance of taxa at some rank (e.g. species, family). Beta diversity also uses the number of relative abundance of taxa at some rank, but measures variation between samples. In other words, an alpha diversity statistic describes a single sample and a beta diversity statistic describes how two samples compare.

Common alpha diversity statistics include Shannon diversity, which measures how difficult it is to predict the identity of a randomly chosen organism inside a sample; and Simpson diversity, which represents the probability that two randomly chosen organisms are the same species.

Beta diversity is a way to quantify the differences between two communities. There are several possible metrics available, the two more commonly used being *Bray–Curtis* beta diversity and *Weighted Unifrac* beta diversity. [**CITATION NEEDED**]. Given that it consists of a series of pair-wise comparisons between samples, beta diversity takes naturally the shape of a triangular matrix. In is thus common to represent beta diversity after doing Multi Dimensional Scaling, so that the two most meaningful dimensions are plotted and cluster could emerge (if present).

All diversity statistics were computed using the R package *vegan* [Oksanen et al., 2019]

References:

* Whittaker, Robert Harding. 1960. “Vegetation of the Siskiyou Mountains, Oregon and California.” Ecological Monographs 30 (3). Wiley Online Library: 279–338.
* Jari Oksanen, F. Guillaume Blanchet, Michael Friendly, Roeland Kindt, Pierre Legendre, Dan McGlinn, Peter R.
  Minchin, R. B. O'Hara, Gavin L. Simpson, Peter Solymos, M. Henry H. Stevens, Eduard Szoecs and Helene Wagner
  (2019). vegan: Community Ecology Package. R package version 2.5-6. https://CRAN.R-project.org/package=vegan


```{r compute alpha, echo=FALSE, include=TRUE}
levels = c('family', 'genus', 'species')

alpha = NULL

for (level in levels){
  #building the OTU table
  OTU_milk = build_OTU_table(raw_OTUs, level) 
  
  #subset to non-control and remove bad sample
  sel = OTU_milk$meta$type != 'ctrl' 
  OTU_milk = filter_OTU(OTU_milk, sel)
  
  alpha = rbind(alpha, data.frame(
    Shannon = diversity(OTU_milk$OTU_noUN, index = 'shannon'),
    Simpson = diversity(OTU_milk$OTU_noUN, index = 'simpson'),
    #invsimpson = diversity(OTU_milk$OTU_noUN, index = 'invsimpson'),
    type = OTU_milk$meta$type, 
    province = OTU_milk$meta$province, 
    sample = rownames(OTU_milk$meta),
    level = level, 
    month = OTU_milk$meta$MESE
  ))
}

#moving to long format, for ggplot
alpha_long = melt(alpha, id.vars = c('sample', 'type', 'level', 'province', 'month'), variable.name = 'alpha')
```

## Alpha diversity - grana vs. non grana
```{r, fig.height = 4, fig.align = "center", echo=FALSE, include=TRUE}
p = ggplot(alpha_long, aes(x=alpha, y=value, fill=type)) + geom_boxplot() +  facet_wrap(~level, nrow = 1) + 
  theme(legend.position = 'bottom', legend.title = element_blank(), axis.title = element_blank()) +
  ggtitle('Alpha diversity: grana vs. non-grana')
print(p)
```

We can formally verify the consistency of this trend via a statistical test. Unfortunately we cannot use the classical Student's T test because the data are not normal (verified via Shapiro test for normality, data not reported). We must thus use the Wilcoxon rank sum test. The next table reports the p-values (alternative hypothesis: diversity are significantly different between grana and non-grana):

```{r, echo=FALSE, include=TRUE}
levels = c('family', 'genus', 'species')
wil = NULL

for (l in levels){
  w_sha = wilcox.test(
    subset(alpha_long, level==l & type=='grana' & alpha == 'Shannon')$value,
    subset(alpha_long, level==l & type=='non-grana' & alpha == 'Shannon')$value
  )
  w_sim = wilcox.test(
    subset(alpha_long, level==l & type=='grana' & alpha == 'Simpson')$value,
    subset(alpha_long, level==l & type=='non-grana' & alpha == 'Simpson')$value
  )
  wil = rbind(wil, data.frame(
    level = l,
    Shannon_pvalue = w_sha$p.value,
    Simpson_pvalue = w_sim$p.value
  ))
  
}

kable(wil)
```

Unfortunately, it appears that there is no statistically significant difference (all p-values are greater than 0.05). This is heavily influenced by one single non-grana sample ([NON-GRANA]_Na_ID1557-15) which has a very high diversity. E.g. for Shannon index at species level we have:

```{r, echo=FALSE, include=TRUE}
tmp = subset(alpha_long, level == 'species' & alpha == 'Shannon' & type == 'non-grana')
kable(tmp)
```

With a Shannon alpha diversity of 1.468 compared to the other three all being lower than 0.5, the "Na" sample is able to drive a lot of the statistical test.

This trend is present also in the "grana" samples. In fact if we take a look at the density distribution (similar to a traditional histogram) of the Shannon alpha diversity we see:

```{r, echo=FALSE, include=TRUE}
p = ggplot(subset(alpha_long, alpha == 'Shannon'), aes(x=value, fill=type)) + geom_density(alpha=0.4) +  facet_wrap(~level, nrow = 1) + 
  theme(legend.position = 'bottom', legend.title = element_blank(), axis.title = element_blank()) +
  ggtitle('Shannon alpha density: grana vs. non-grana')
print(p)
```

It appears clear that:

* grana samples are by far non-normal. In fact they seem to be bimodal: samples with high or low diversity, with little middle ground
* this bimodal trend is present in all taxonomic levels
* same results with Simpson diversity, not reported
* at least at species level, the non-grana samples follow the same distribution, as previously noted

The bimodal trend is actually intrinsic to the data. Even samples from the same dairy show it, e.g. for milk source "CN911":

```{r, echo=FALSE, include=TRUE}
tmp = subset(alpha_long, grepl(alpha_long$sample, pattern = 'CN911') & alpha == 'Shannon' & level == 'species')
kable(tmp, row.names = FALSE)
```

Or milk source "VI611":

```{r, echo=FALSE, include=TRUE}
tmp = subset(alpha_long, grepl(alpha_long$sample, pattern = 'VI611') & alpha == 'Shannon' & level == 'species')
kable(tmp, row.names = FALSE)
```

### Open question: are diary sources uniform?

When I see data coming from a single diary (e.g. CN911) does it mean that the milk was produced **in the same farm**?

## Alpha diversity, provinces
```{r, fig.height = 4, fig.align = "center", echo=FALSE, include=TRUE}
p = ggplot(subset(alpha_long, province != '-'), aes(x=alpha, y=value, fill=province)) + geom_boxplot() +  facet_wrap(~level, nrow = 1) + 
  theme(legend.position = 'bottom', axis.title = element_blank()) +
  scale_fill_discrete(guide = guide_legend(nrow = 1, title = NULL) ) +
  ggtitle('Alpha diversity: provinces')
print(p)
```

Given that we have multiple groups (i.e. multiple provinces) we can perform an analysis of variance (ANOVA), which will tell if there are any statistically significant differences between the groups. The p-values are reported in the following table:

```{r alpha ANOVA, echo=FALSE, include=TRUE, eval=TRUE}
levels = c('family', 'genus', 'species')
ano = NULL

for (l in levels){
  df_sha = subset(alpha_long, level==l & type=='grana' & alpha == 'Shannon')
  a_sha = aov(value ~ province, data = df_sha)
  s_sha = summary(a_sha)
  df_sim = subset(alpha_long, level==l & type=='grana' & alpha == 'Simpson')
  a_sim = aov(value ~ province, data = df_sim)
  s_sim = summary(a_sim)
  
  ano = rbind(ano, data.frame(
    level = l,
    Shannon_pvalue = s_sha[[1]][1,'Pr(>F)'],
    Simpson_pvalue = s_sim[[1]][1,'Pr(>F)']
  ))
}

kable(ano)
```

It appears clear that the deeper we go into the taxonomic tree the more significant the differences between provinces are. Moreover, Simpson test gives, in general, lower p-values, meaning that this metric is able to better intercept and summarize the existing differences among our groups.

## Rarefaction curves

The following plot shows the rarefaction curves, where each line represents a sample. This plot links the amount of detected species as a function of the number of reads. Intuitively, if the reads are too few not all species will be detected. But after a certain point more reads are not useful because no more species are detected. Labels for samples are not reported to avoid plot overcrowding, but could be put back if required.

```{r rarefaction curves, echo=FALSE, eval=TRUE}
level = 'species'

#building the OTU table
OTU_milk = build_OTU_table(raw_OTUs, level) 

#subset to non-control and remove bad sample
sel = OTU_milk$meta$type != 'ctrl' 
OTU_milk = filter_OTU(OTU_milk, sel)

min_reads = min(rowSums(OTU_milk$OTU_noUN))
cur = rarecurve(OTU_milk$OTU_noUN, step = 1, col = "blue", label = FALSE, cex = 0.6)
abline(v=min_reads)
```

The plot is a bit problematic. Some lines take a very long time to reach saturation, with the slowest one requiring ~7000 reads before becoming horizontal, and a couple of other lines requiring ~3000 reads. Unfortunately, nine lines stop before 1000 reads and twenty before 2000, as reported in the following table (note: we are talking about assigned reads, i.e. we are not counting unassigned, probably bovine reads):

```{r rarefaction stats, echo=FALSE, eval=TRUE}
binned_reads = data.frame(
  threshold = c(250, 500, 750, 1000, 1500, 2000, 4000, 6000, 13000)
)
x = rowSums(OTU_milk$OTU_noUN)

for (i in 1:nrow(binned_reads)){
  binned_reads[i, 'samples'] = sum(x < binned_reads[i, 'threshold'])
}

kable(binned_reads)
```

The conclusion is that a non negligible fraction of the samples (a third?) could have used more reads.

## Beta diversity

I've introduced the concept of beta diversity at the beginning of this section. For simplicity all shown data are at "species" level, and using the "Bray-Curtis" metric. I found little difference using other parameters, however.


```{r compute beta, echo=FALSE, include=FALSE, eval=TRUE}
level = 'species'

#building the OTU table
OTU_milk = build_OTU_table(raw_OTUs, level) 

#subset to non-control and remove bad sample
sel = OTU_milk$meta$type != 'ctrl' 
OTU_milk = filter_OTU(OTU_milk, sel)

#building the beta distribution matrix, then MDS
beta_dist = vegdist(OTU_milk$OTU_noUN, index = "bray")  
mds <- metaMDS(beta_dist, try = 100, trymax = 100, trace = 0)

#building a usable dataframe
mds_data <- as.data.frame(mds$points)
mds_data$code <- rownames(mds_data)
mds_data <- dplyr::left_join(mds_data, OTU_milk$meta, by = "code")
mds_data$ccf = grepl(mds_data$Id_Lab, pattern = 'ccf')

#custom shapes
shapes_prov = c(BS=17,  CN=18, CR=19, PC=0, TN=1, VI=2, VR=3)
```

### By type, grana vs. non grana
```{r beta by type, echo=FALSE, include=TRUE, eval=TRUE}
#building a column for labels
mds_data$labs = mds_data$Id_Lab
mds_data[mds_data$type == 'grana', 'labs'] = NA

ggplot(mds_data, aes(x = MDS1, y = MDS2, color = type, label=labs)) + geom_point() + geom_label() + theme(legend.position = 'none')
```

In this plot each milk sample from the grana consortium is represented with a red dot. The four non-grana milk are reported as blue labels. A few considerations:

* the milk samples are highly divided in two clusters, amplifying the general structure already seen in alpha diversity analysis
* the non-grana "Na" sample is separated from the other three, as already suspected in the alpha diversity analysis

### By province
```{r beta by prov, echo=FALSE, include=TRUE, eval=TRUE}
ggplot(subset(mds_data, province!= '-'), aes(x = MDS1, y = MDS2, color = province, shape=province)) +
  scale_shape_manual(values = shapes_prov) +
  geom_point()
```

The data are exactly the same as in the previous plot, but here I highlight the geographical source. Unfortunately, not many patterns are visible.

### By ccf

I plot again the same data, but highlighting extraction method - it was very easy to do :)

```{r beta by ccf, echo=FALSE, include=TRUE, eval=TRUE}
ggplot(mds_data, aes(x = MDS1, y = MDS2, color = ccf)) + geom_point()
```

### By sampling date

I've investigated the sampling date, too, to see if it could explain the visible clusters, with little succes. Data not reported for brevity.

# Taxon abundance 
[TODO]
```{r, fig.height = 4, fig.align = "center", echo=FALSE, include=TRUE}
#see taxa in roots, here https://grunwaldlab.github.io/analysis_of_microbiome_community_data_in_r/07--diversity_stats.html
# obj$data$type_abund <- calc_taxon_abund(obj, "otu_rarefied",
#                                         cols = sample_data$SampleID,
#                                         groups = sample_data$Type)

```


